src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs

```diff
@@ -66,61 +68,64 @@ public override void Read(JsonTokenType tokenType, JsonSerializerOptions options
         // If this method is changed, also change JsonPropertyInfoNullable.ReadEnumerable and JsonSerializer.ApplyObjectToEnumerable
         public override void ReadEnumerable(JsonTokenType tokenType, JsonSerializerOptions options, ref ReadStack state, ref Utf8JsonReader reader)
         {
+            Debug.Assert(ShouldDeserialize);
+
             if (ValueConverter == null || !ValueConverter.TryRead(RuntimePropertyType, ref reader, out TRuntimeProperty value))
             {
                 ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(RuntimePropertyType, reader, state.PropertyPath);
                 return;
             }
 
-            JsonSerializer.ApplyValueToEnumerable(ref value, options, ref state, ref reader);
+            JsonSerializer.ApplyValueToEnumerable(ref value, ref state, ref reader);
         }
 
         public override void Write(JsonSerializerOptions options, ref WriteStackFrame current, Utf8JsonWriter writer)
         {
             Debug.Assert(current.Enumerator == null);
+            Debug.Assert(ShouldSerialize);
 
-            if (ShouldSerialize)
+            TRuntimeProperty value;
+            if (IsPropertyPolicy)
             {
-                TRuntimeProperty value;
-                if (_isPropertyPolicy)
-                {
-                    value = (TRuntimeProperty)current.CurrentValue;
-                }
-                else
+                value = (TRuntimeProperty)current.CurrentValue;
+            }
+            else
+            {
+                value = (TRuntimeProperty)Get((TClass)current.CurrentValue);
+            }
+
+            if (value == null)
+            {
+                Debug.Assert(_escapedName != null);
```

Comment:
Why is this true? What if the user actually wants to write the null value within a JSON array (without a property name)?

Reply:
Null values within an array work and are handled higher up the stack.

However, a null property name doesn't work today. I don't think the reader supports it although I believe it is valid JSON.

Reply:
> However, a null property name doesn't work today. I don't think the reader supports it although I believe it is valid JSON.

The string with the contents `n`, `u`, `l`, `l` within quotes is valid JSON (e.g. `{"null": "foo"}`), but I don't think just the null literal is a valid property name. What should the reader be supporting here?